File: V1/mac_changer.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <time.h>

#define INTERFACE "wlan0"
#define CHANGE_INTERVAL 1800 // 30 minutes
#define COUNTDOWN_STEP 300   // 5 minutes

void get_current_mac(char *interface, char *mac_address) {
    FILE *fp;
    char cmd[100];
    snprintf(cmd, sizeof(cmd), "cat /sys/class/net/%s/address", interface);
    
    fp = popen(cmd, "r");
    if (fp == NULL) {
        perror("Failed to get MAC address");
        exit(1);
    }
    
    fgets(mac_address, 18, fp);
    pclose(fp);
}

void generate_random_mac(char *mac_address) {
    srand(time(NULL));
    snprintf(mac_address, 18, "%02X:%02X:%02X:%02X:%02X:%02X",
             rand() % 256, rand() % 256, rand() % 256,
             rand() % 256, rand() % 256, rand() % 256);
}

void change_mac(char *interface) {
    char old_mac[18], new_mac[18], cmd[200];
    
    get_current_mac(interface, old_mac);
    generate_random_mac(new_mac);
    
    printf("[!] Changing MAC address from %s to %s\n", old_mac, new_mac);
    
    snprintf(cmd, sizeof(cmd), "sudo ifconfig %s down", interface);
    system(cmd);
    
    snprintf(cmd, sizeof(cmd), "sudo macchanger -m %s %s", new_mac, interface);
    system(cmd);
    
    snprintf(cmd, sizeof(cmd), "sudo ifconfig %s up", interface);
    system(cmd);
    
    char updated_mac[18];
    get_current_mac(interface, updated_mac);
    
    if (strcmp(new_mac, updated_mac) == 0) {
        printf("[+] MAC successfully changed to %s\n", updated_mac);
    } else {
        printf("[-] MAC change failed!\n");
    }
}

void countdown_timer(char *interface) {
    while (1) {
        for (int remaining = CHANGE_INTERVAL; remaining > 0; remaining -= COUNTDOWN_STEP) {
            printf("Next MAC change in %d minutes...\n", remaining / 60);
            sleep(COUNTDOWN_STEP);
        }
        change_mac(interface);
    }
}

int main() {
    printf("\n[+] MAC Changer Script Starting...\n");
    printf("[!] Interface: %s\n", INTERFACE);
    countdown_timer(INTERFACE);
    return 0;
}

File: V1/mac_changer.py
import time
import random
import re
import subprocess

def get_current_mac(interface="eth0"):
    """Get the current MAC address of the given interface."""
    try:
        output = subprocess.check_output(f"ifconfig {interface}", shell=True).decode()
        mac_address = re.search(r"(\w\w:\w\w:\w\w:\w\w:\w\w:\w\w)", output)
        if mac_address:
            return mac_address.group(0)
        return None
    except Exception as e:
        print(f"Error getting MAC address: {e}")
        return None

def generate_random_mac():
    """Generate a random MAC address."""
    return ":".join(f"{random.randint(0, 255):02x}" for _ in range(6))

def change_mac(interface="eth0"):
    """Change the MAC address of the given interface."""
    old_mac = get_current_mac(interface)
    new_mac = generate_random_mac()
    
    try:
        print(f"Changing MAC address from {old_mac} to {new_mac}")
        subprocess.call(f"sudo ifconfig {interface} down", shell=True)
        subprocess.call(f"sudo ifconfig {interface} hw ether {new_mac}", shell=True)
        subprocess.call(f"sudo ifconfig {interface} up", shell=True)
        
        updated_mac = get_current_mac(interface)
        
        if updated_mac == new_mac:
            print(f"MAC successfully changed to {updated_mac}")
        else:
            print("MAC change failed!")
    except Exception as e:
        print(f"Error changing MAC address: {e}")

def countdown_timer(interval=1800, echo_interval=300, interface="eth0"):
    """Run change_mac() every `interval` seconds while echoing every `echo_interval` seconds."""
    while True:
        for remaining in range(interval, 0, -echo_interval):
            print(f"Next MAC change in {remaining // 60} minutes...")
            time.sleep(echo_interval)
        change_mac(interface)

if __name__ == "__main__":
    countdown_timer()